File: frontend/src/api/task.ts
import request from '../utils/request'

export interface Task {
  id: number
  title: string
  description?: string
  status: 'wait' | 'doing' | 'done' | 'pause' | 'cancel' | 'closed'
  priority: 'low' | 'medium' | 'high' | 'urgent'
  project_id: number
  project?: any
  requirement_id?: number
  requirement?: any
  creator_id: number
  creator?: any
  assignee_id?: number
  assignee?: any
  start_date?: string
  end_date?: string
  due_date?: string
  progress: number
  estimated_hours?: number
  actual_hours?: number
  dependencies?: Task[]
  created_at?: string
  updated_at?: string
}

export interface TaskListResponse {
  list: Task[]
  total: number
  page: number
  page_size: number
}

export interface CreateTaskRequest {
  title: string
  description?: string
  status?: 'wait' | 'doing' | 'done' | 'pause' | 'cancel' | 'closed'
  priority?: 'low' | 'medium' | 'high' | 'urgent'
  project_id: number
  requirement_id?: number
  assignee_id?: number
  start_date?: string
  end_date?: string
  due_date?: string
  progress?: number
  estimated_hours?: number
  actual_hours?: number
  work_date?: string
  dependency_ids?: number[]
}

export interface UpdateTaskStatusRequest {
  status: 'wait' | 'doing' | 'done' | 'pause' | 'cancel' | 'closed'
}

export interface UpdateTaskProgressRequest {
  progress?: number
  estimated_hours?: number
  actual_hours?: number
  work_date?: string
}

// 任务相关API
export const getTasks = async (params?: {
  keyword?: string
  project_id?: number
  status?: string
  priority?: string
  assignee_id?: number
  creator_id?: number
  page?: number
  size?: number
}): Promise<TaskListResponse> => {
  return request.get('/tasks', { params })
}

export const getTask = async (id: number): Promise<Task> => {
  return request.get(`/tasks/${id}`)
}

export const createTask = async (data: CreateTaskRequest): Promise<Task> => {
  return request.post('/tasks', data)
}

export const updateTask = async (id: number, data: Partial<CreateTaskRequest>): Promise<Task> => {
  return request.put(`/tasks/${id}`, data)
}

export const deleteTask = async (id: number): Promise<void> => {
  return request.delete(`/tasks/${id}`)
}

export const updateTaskStatus = async (id: number, data: UpdateTaskStatusRequest): Promise<Task> => {
  return request.patch(`/tasks/${id}/status`, data)
}

export const updateTaskProgress = async (id: number, data: UpdateTaskProgressRequest): Promise<Task> => {
  return request.patch(`/tasks/${id}/progress`, data)
}

// 历史记录相关接口（参考bug.ts、project.ts和requirement.ts）
export interface History {
  id: number
  action_id: number
  field: string
  old: string
  old_value: string
  new: string
  new_value: string
  diff?: string
  created_at?: string
}

export interface Action {
  id: number
  object_type: string
  object_id: number
  project_id: number
  actor_id: number
  actor?: {
    id: number
    username: string
    nickname?: string
  }
  action: 'created' | 'edited' | 'assigned' | 'resolved' | 'closed' | 'confirmed' | 'commented' | 'status_changed' | 'progress_updated'
  date: string
  comment?: string
  extra?: string
  histories?: History[]
  created_at?: string
}

export interface TaskHistoryResponse {
  list: Action[]
}

export interface AddTaskHistoryNoteRequest {
  comment: string
}

export const getTaskHistory = async (id: number): Promise<TaskHistoryResponse> => {
  return request.get(`/tasks/${id}/history`)
}

export const addTaskHistoryNote = async (id: number, data: AddTaskHistoryNoteRequest): Promise<{ message: string }> => {
  return request.post(`/tasks/${id}/history/note`, data)
}

File: backend/internal/model/task.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// Task 任务表
type Task struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	Title       string `gorm:"size:200;not null" json:"title"`        // 任务标题
	Description string `gorm:"type:text" json:"description"`         // 任务描述（Markdown）
	Status      string `gorm:"size:20;default:'todo'" json:"status"`  // 状态：todo, in_progress, done, cancelled
	Priority    string `gorm:"size:20;default:'medium'" json:"priority"` // 优先级：low, medium, high, urgent

	ProjectID uint    `gorm:"index;not null" json:"project_id"`
	Project   Project `gorm:"foreignKey:ProjectID" json:"project,omitempty"`

	RequirementID *uint       `gorm:"index" json:"requirement_id"`
	Requirement   *Requirement `gorm:"foreignKey:RequirementID" json:"requirement,omitempty"`

	CreatorID uint `gorm:"index" json:"creator_id"`
	Creator   User `gorm:"foreignKey:CreatorID" json:"creator,omitempty"`

	AssigneeID *uint `gorm:"index" json:"assignee_id"`
	Assignee   *User `gorm:"foreignKey:AssigneeID" json:"assignee,omitempty"`

	StartDate *time.Time `json:"start_date"` // 开始日期
	EndDate   *time.Time `json:"end_date"`   // 结束日期
	DueDate   *time.Time `json:"due_date"`   // 截止日期

	Progress int `gorm:"default:0" json:"progress"` // 进度：0-100

	EstimatedHours *float64 `gorm:"default:0" json:"estimated_hours"` // 预估工时（小时）
	ActualHours    *float64 `gorm:"default:0" json:"actual_hours"`    // 实际工时（小时），从资源分配自动计算

	// 任务依赖关系（多对多）
	Dependencies []Task `gorm:"many2many:task_dependencies;joinForeignKey:task_id;joinReferences:dependency_id" json:"dependencies,omitempty"`
}

// TaskDependency 任务依赖关系表
type TaskDependency struct {
	TaskID       uint `gorm:"primaryKey" json:"task_id"`
	DependencyID uint `gorm:"primaryKey" json:"dependency_id"`
	Type         string `gorm:"size:20;default:'finish_to_start'" json:"type"` // 依赖类型：finish_to_start, start_to_start, finish_to_finish, start_to_finish
}

// Board 看板表
type Board struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	Name        string `gorm:"size:100;not null" json:"name"`        // 看板名称
	Description string `gorm:"type:text" json:"description"`            // 描述

	ProjectID uint    `gorm:"index;not null" json:"project_id"`
	Project   Project `gorm:"foreignKey:ProjectID" json:"project,omitempty"`

	Columns []BoardColumn `gorm:"foreignKey:BoardID" json:"columns,omitempty"`
}

// BoardColumn 看板列表
type BoardColumn struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	Name  string `gorm:"size:50;not null" json:"name"`  // 列名称
	Color string `gorm:"size:20" json:"color"`           // 列颜色
	Sort  int    `gorm:"default:0" json:"sort"`          // 排序

	BoardID uint   `gorm:"index;not null" json:"board_id"`
	Board   Board  `gorm:"foreignKey:BoardID" json:"board,omitempty"`

	Status string `gorm:"size:20" json:"status"` // 关联的任务状态
}

